### PR TITLE
QGIS 3.2 Port

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR ports Inasafe to QGIS 3.2/Python 3/Qt 5.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@ismailsunni 

When you are testing, can you please look at https://github.com/inasafe/inasafe/pull/5000/commits/403772c5f979bd35b914b0915cc1e29c7589da9c (reverted in https://github.com/inasafe/inasafe/pull/5000/commits/394de29de50e4a9e9f7596238967af9889590d60 )?

I suspect this is a real bug in Inasafe, but fixing it breaks MANY of the existing unit test expected results. Happy to discuss via gitter/etc .

**Comment by Gustry** [comment]
Thanks a lot @elpaso and @nyalldawson. I will try to check check this PR and these 2 commits too.

With a quick look, it doesn't seem to be a strict port to python3, qt5 and qgis3 API. But you also made some improvements. Is-it possible to have a quick list where you updated/improved the code to embrace better the QGIS 3 API?
For instance, did you drop our own code about union/intersections/multibuffer algs to use the native ones? (especially the work from Martin about porting to C++ these algorithms in QGIS 3.2).

**Comment by nyalldawson** [comment]
@Gustry

It's mostly a direct port. There's been a bit of refactoring in the report class, but that's mostly due to the large API difference between layouts and 2.x composer.

There's also some smaller optimisations (e.g. https://github.com/inasafe/inasafe/pull/5000/commits/e839b89e2136f2fa6edf3891cc1c4546ae4ea728 ), but I'd prefer to leave these optimisations for follow up work. 

My thoughts are:

- We could easily replace clip, intersection, reproject and union by the native processing algorithm without any risk or loss of functionality (and probably much improved performance + robustness). 

- There's no direct replacement for multi_buffering in processing (yet) - there's the `multi-ring buffer (constant distance)` algorithm, but that requires a constant buffer step size. So we'd need first to implement a native multi-ring buffer algorithm which accepts a list of values to buffer by.

- smart_clip is another candidate for replacement, but I'm unsure what the logic is here and why this is sometimes used instead of clip. Any ideas?

- contour could possibly be replaced by the gdal contour alg, but there's a lot happening in the inasafe contour code which I'm not sure maps across to gdal functionality

- reclassify could potentially be replaced by https://github.com/qgis/QGIS/pull/7203, but we'll need to see if that work lands for 3.2

File: safe/common/parameters/group_select_parameter_widget.py [hunk 1]
```diff
@@ -3,16 +3,17 @@
...
-    QVBoxLayout, QRadioButton, QButtonGroup, QWidget, QLabel, QSizePolicy,
-    QListWidget, QGridLayout, QAbstractItemView, QListWidgetItem)
-
-from parameters.qt_widgets.generic_parameter_widget import (
-    GenericParameterWidget)
-from safe.common.parameters.percentage_parameter_widget import (
-    PercentageSpinBox)
-from safe.definitions.constants import STATIC, SINGLE_DYNAMIC, MULTIPLE_DYNAMIC
+from qgis.PyQt.QtCore import Qt
+from qgis.PyQt.QtWidgets import (QAbstractItemView, QButtonGroup, QGridLayout,
```
Comment (timlinux) [inline safe/common/parameters/group_select_parameter_widget.py]: Pull left -see  https://github.com/inasafe/inasafe/wiki/Coding-Standards#code-formatting

**Comment by timlinux** [comment]
Thank you so much for this PR. For merging could you please:

* reenable E501 checks in pyflake8 so that we can flag long lines
* get all lines 'pull left' as per https://github.com/inasafe/inasafe/wiki/Coding-Standards#code-formatting
* get all lines < 80 chars

Then we can merge and continue with smaller, more managable PRs

**Comment by elpaso** [comment]
@timlinux flake8 should be now ok in https://github.com/north-road/inasafe/pull/44

**Comment by timlinux** [comment]
> @timlinux flake8 should be now ok in north-road#44

Thanks @elpaso 

@ismailsunni As soon as the tests (not all the slider warnings if they are about duplicated code), lets get their stuff merged.

Also @ismailsunni I am thinking for the QGIS 3 port we should maybe drop the v3 keywords stuff to keep things clean - what do you think?

**Comment by ismailsunni** [comment]
Hi @timlinux 
> @ismailsunni As soon as the tests pass (not all the slider warnings if they are about duplicated code), lets get their stuff merged.

Yeah, I agree.

> Also @ismailsunni I am thinking for the QGIS 3 port we should maybe drop the v3 keywords stuff to keep things clean - what do you think?

Well, we create it for GeoSAFE, since we don't have (yet) GeoSAFE that supports v4, I am not sure how the user will do it from QGIS 3. (You can do it in InaSAFE in QGIS2.18, so maybe it's fine)

File: safe/common/parameters/group_select_parameter_widget.py (hunk 1 above)
Comment (nyalldawson) [reply safe/common/parameters/group_select_parameter_widget.py]: @elpaso can you please scan through and fix any others like this too?

File: .gitmodules [hunk 2]
```diff
@@ -1,3 +1,3 @@
 [submodule "safe_extras/parameters"]
 	path = safe_extras/parameters
-	url = https://github.com/inasafe/parameters.git
+	url = https://github.com/north-road/parameters.git
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply .gitmodules]: @elpaso can you please fix?

File: safe/report/test/test_impact_report.py [hunk 3]
```diff
@@ -7,64 +7,57 @@
...
+from safe.definitions.constants import ANALYSIS_SUCCESS, INASAFE_TEST, PREPARE_SUCCESS
+from safe.definitions.field_groups import (age_displaced_count_group,
+                                           gender_displaced_count_group)
+from safe.definitions.fields import (total_affected_field, total_exposed_field,
+                                     total_not_affected_field,
+                                     total_not_exposed_field)
 from safe.definitions.hazard import hazard_flood
+from safe.definitions.hazard_classifications import flood_hazard_classes
 from safe.definitions.provenance import provenance_use_rounding
+from safe.definitions.reports.components import (action_checklist_component,
```
Comment (ismailsunni) [inline safe/report/test/test_impact_report.py]: Pull left

File: safe/report/test/test_impact_report.py [hunk 4]
```diff
@@ -7,64 +7,57 @@
...
-    minimum_needs_component,
-    aggregation_postprocessors_component,
-    population_infographic_component,
-    analysis_provenance_details_component,
-    analysis_provenance_details_simplified_component,
-    standard_multi_exposure_impact_report_metadata_html)
-from safe.definitions.utilities import update_template_component
 from safe.report.impact_report import ImpactReport
+from safe.report.report_metadata import ReportMetadata
+from safe.test.utilities import (get_qgis_app, load_test_raster_layer,
```
Comment (ismailsunni) [inline safe/report/test/test_impact_report.py]: Pull left and some more in this file

File: safe/test/utilities.py [hunk 5]
```diff
@@ -9,28 +10,30 @@
...
-from PyQt4 import QtGui  # pylint: disable=W0621
-from PyQt4.QtCore import QTranslator, pyqtWrapperType
-from qgis.core import (
-    QgsVectorLayer,
-    QgsRasterLayer,
-    QgsRectangle,
-    QgsCoordinateReferenceSystem,
-    QgsMapLayerRegistry)
+import processing
+from qgis.core import (QgsCoordinateReferenceSystem, QgsProject,
```
Comment (ismailsunni) [inline safe/test/utilities.py]: Pull left

File: safe/test/utilities.py [hunk 6]
```diff
@@ -1177,8 +1182,26 @@ def get_control_text(file_name):
...
+            mode='r',
+            encoding='utf-8') as f:
+        return f.readlines()
+    return ''
+
+
+def dict_values_sorted(d):
+    """
+    Make sure dict values are sorted when they are sortable.
+    This also works for lists of dicts nd discts of lists
```
Comment (ismailsunni) [inline safe/test/utilities.py]: `nd discts` ?
Can you give the `param` and the `returns`?

File: safe/utilities/extent.py [hunk 7]
```diff
@@ -2,21 +2,14 @@
...
-    last_analysis_color,
-    user_analysis_width,
-    next_analysis_width,
-    last_analysis_width
-)
+from qgis.core import (QgsCoordinateTransform, QgsGeometry, QgsProject,
+                       QgsWkbTypes)
+from qgis.gui import QgsRubberBand
+from qgis.PyQt.QtCore import Qt
+from safe.definitions.styles import (last_analysis_color, last_analysis_width,
```
Comment (ismailsunni) [inline safe/utilities/extent.py]: Pull left

File: safe/utilities/help.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    html_footer,
+    html_help_header,
+)
+
+
+def get_help_html(message=None):
+    """Create the HTML content for the help dialog  or for external browser
+
+    :param message: An optional message object to display in the dialog.
+    :type message: Message.Message
```
Comment (ismailsunni) [inline safe/utilities/help.py]: New line after to separate param and returns

File: safe/utilities/settings.py [hunk 9]
```diff
@@ -16,6 +17,29 @@
 __revision__ = '$Format:%H$'
 
 
+def deep_convert_dict(value):
+    """
+    Converts any OrderedDict elements in a value to
```
Comment (ismailsunni) [inline safe/utilities/settings.py]: Short sentence in the first line to describe the method

**Comment by ismailsunni** [comment]
Hi @nyalldawson and @elpaso 
Thanks, I will merge this to develop branch.

**Comment by elpaso** [comment]
@ismailsunni thank **you** for the review and for all your help along the process!

**Pull request merged**